### PR TITLE
[feat](Nereids) Support tablesample syntax

### PR DESCRIPTION
**Pull request by Kikyou1997**
## Proposed changes

Add table sample support like such SQL:

```sql
select * from test_table_sample_tbl tablesample(4 rows);
select * from test_table_sample_tbl t tablesample(20 percent);
select * from test_table_sample_tbl t tablesample(20 percent) repeatable 2;
```

This function has already been implemented in legacy planner, more detailed description could be found here: https://github.com/apache/doris/pull/10170

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.18 seconds
 stream load tsv:          551 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162016548 Bytes

**Review by englefly — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java [hunk 1]
```diff
@@ -44,6 +45,7 @@
  */
 public class UnboundRelation extends LogicalRelation implements Unbound {
 
+    public final TableSample tableSample;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java:48]: use Optional to avoid use null directly

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/TableSample.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ *      TABLESAMPLE (10 PERCENT) REPEATABLE 2
+ *      TABLESAMPLE (100 ROWS) REPEATABLE 2
+ */
+public class TableSample {
+
+    public final Long sampleValue;
+
+    public final boolean isPercent;
+
+    public Long seek = -1L;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/TableSample.java:35]: why use Long, not long?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/TableSample.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * To represent following stmt:
+ *      TABLESAMPLE (10 PERCENT)
+ *      TABLESAMPLE (100 ROWS)
+ *      TABLESAMPLE (10 PERCENT) REPEATABLE 2
+ *      TABLESAMPLE (100 ROWS) REPEATABLE 2
+ */
+public class TableSample {
+
+    public final Long sampleValue;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/TableSample.java:31]: why use Long, not long?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/TableSample.java (hunk 2 above)
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/TableSample.java:35]: final

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/TableSample.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.Objects;
+
+/**
+ * To represent following stmt:
+ *      TABLESAMPLE (10 PERCENT)
+ *      TABLESAMPLE (100 ROWS)
+ *      TABLESAMPLE (10 PERCENT) REPEATABLE 2
+ *      TABLESAMPLE (100 ROWS) REPEATABLE 2
+ */
+public class TableSample {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/TableSample.java:29]: TableSample is not an expression, put it under trees

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java [hunk 5]
```diff
@@ -51,6 +52,8 @@
  */
 public class LogicalOlapScan extends LogicalCatalogRelation implements OlapScan {
 
+    public final TableSample tableSample;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:55]: Optional. put it as last var of this class

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java [hunk 6]
```diff
@@ -45,6 +46,7 @@
  */
 public class PhysicalOlapScan extends PhysicalCatalogRelation implements OlapScan {
 
+    public final TableSample tableSample;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java:49]: Optional, put it under baseOutputs

Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java:49]: It's not allowed by our checkstyle rules

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java (hunk 5 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:55]: Declare public fields after privete is not allowed by our checkstyle rules

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java (hunk 1 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java:48]: According to the one who proposed Optional api to Java8, seems we should almost never wrap a member with Optional, the only scenario use this API is the return type of a function https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.12 seconds
 stream load tsv:          528 seconds loaded 74807831229 Bytes, about 135 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162102113 Bytes

**Comment by Kikyou1997** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.91 seconds
 stream load tsv:          536 seconds loaded 74807831229 Bytes, about 133 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162084732 Bytes

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java (hunk 1 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java:48]: I checked the codes by executing command `grep -r 'private Optional' .` under org/apache/doris/nereids package, there's only few fields declared as `Optional<SomeFields> aFields;`

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java (hunk 1 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java:48]: if u insist on using null, i cannot merge this PR. because NPE is hard to follow and the error message is useless for debuging

**Review by morrySnow — CHANGES_REQUESTED** [review]

**Comment by Kikyou1997** [comment]
run buildall

**Comment by englefly** [comment]
could you add some explanation about the above sqls?
is "tablesample " keyword? 
what is "repeatable" in the last sql?

**Comment by Kikyou1997** [comment]
> could you add some explanation about the above sqls? is "tablesample " keyword? what is "repeatable" in the last sql?

This function is already implemented by legacy optimizer, you should find this informatioi

**Comment by Kikyou1997** [comment]
> could you add some explanation about the above sqls? is "tablesample " keyword? what is "repeatable" in the last sql?


This function has already been implemented in legacy planner, more detailed description could be found here: https://github.com/apache/doris/pull/10170

**Review by englefly — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/load/ExportJob.java [hunk 7]
```diff
@@ -372,7 +372,7 @@ private LogicalPlan generateOneLogicalPlan(List<String> qualifiedTableName, List
             List<String> partitions, List<NamedExpression> selectLists) {
         // UnboundRelation
         LogicalPlan plan = new UnboundRelation(StatementScopeIdGenerator.newRelationId(), qualifiedTableName,
-                partitions, false, tabletIds, ImmutableList.of());
+                partitions, false, tabletIds, ImmutableList.of(), null);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/load/ExportJob.java:375]:
```suggestion
                partitions, false, tabletIds, ImmutableList.of(), Optional.empty());
```
?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java [hunk 8]
```diff
@@ -44,6 +45,7 @@
  */
 public class UnboundRelation extends LogicalRelation implements Unbound {
 
+    public final Optional<TableSample> tableSample;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java:48]: why public? make it private and add getter

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java (hunk 6 above)
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java:49]: optional private

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Review by morrySnow — DISMISSED** [review]

**Comment by Kikyou1997** [comment]
run buildall

**Review by englefly — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]
